crates/ibc/src/core/ics24_host/validate.rs

```diff
@@ -48,43 +50,100 @@ pub fn validate_identifier(id: &str, min: usize, max: usize) -> Result<(), Error
     Ok(())
 }
 
+/// Checks if the prefix is valid and can form valid domain identifiers.
+pub fn validate_prefix_identifier(
+    prefix: &str,
+    min_id_length: usize,
+    max_id_length: usize,
+) -> Result<(), Error> {
+    // Checks if the prefix is not blank
+    if prefix.is_empty() {
+        return Err(Error::Empty)?;
+    }
+
+    // Checks of the prefix does not end with a dash
+    let re = safe_regex::regex!(br".*[^-]");
+    if !re.is_match(prefix.as_bytes()) {
+        return Err(Error::InvalidPrefix {
+            prefix: prefix.to_string(),
+        })?;
+    }
+
+    // Checks if the prefix forms a valid default identifier when used with `0`
+    validate_default_identifier(
+        &format!("{prefix}-{}", u64::MIN),
+        min_id_length,
+        max_id_length,
+    )?;
+
+    // Checks if the prefix forms a valid default identifier when used with `u64::MAX`
+    validate_default_identifier(
+        &format!("{prefix}-{}", u64::MAX),
+        min_id_length,
+        max_id_length,
+    )?;
+
+    Ok(())
+}
+
+/// Checks if the identifier is of valid format and can be parsed into the
+/// correct domain identifier type.
+pub fn validate_identifier(id: &str, min: usize, max: usize) -> Result<(), Error> {
+    let split_id: Vec<_> = id.split('-').collect();
+    let last_index = split_id.len() - 1;
+    let prefix = split_id[..last_index].join("-");
+
+    validate_prefix_identifier(prefix.trim(), min, max)?;
```

Comment:
The whole identifier must be between min/max length; not just the prefix

Reply:
Ah, the `validate_prefix_identifier` does the length checks with the "full id" using 0 and `u64::MAX`. Although correct, that's a little confusing.

A more general question: what was wrong with the previous identifier checks?

Reply:
It is so.
`validate_prefix_identifier` performs this check just after creating "min/max default identifiers" and passing them to the `validate_default_identifier`

Reply:
Before, [here](https://github.com/cosmos/ibc-rs/blob/1396ad6c82c454444114461f7f55c4509a4727a5/crates/ibc/src/core/ics24_host/identifier.rs#L277-L283) we were allowing wrong `ClientId` creation. E.g. "mock_clientid", "nonexistingclient" were ok.
Also, "07-tendermint-" and "  07-tendermint " as a valid `ClientType`s were accepted.

Reply:
They are okay though; according to the spec, those are all valid identifiers, and this is the only requirement for client identifiers. Our `ChainId` type also [properly handles](https://github.com/cosmos/ibc-rs/blob/4af0a70fe860b6b08555c54970cb69c1a209305f/crates/ibc/src/core/ics24_host/identifier.rs#L70-L74) identifiers that are not in the "epoch" format

Reply:
I believe ibc-go has these extra restrictions, but [only for their `ClientType`](https://github.com/cosmos/ibc-go/blob/295bab01f633633e0d6136ba9b974c0e248edbe5/modules/core/02-client/types/msgs.go#L68). Other chains are free to name their clients without these extra restrictions

Reply:
> They are okay though; according to the spec, those are all valid identifiers, and this is the only requirement for client identifiers. Our ChainId type also [properly handles](https://github.com/cosmos/ibc-rs/blob/4af0a70fe860b6b08555c54970cb69c1a209305f/crates/ibc/src/core/ics24_host/identifier.rs#L70-L74) identifiers that are not in the "epoch" format

Our client creator restricts. https://github.com/cosmos/ibc-rs/pull/639#discussion_r1181623161
Thus, Should we move away from the [client counter](https://github.com/cosmos/ibc-rs/blob/5f02452fa533804b769c42b91e452425ee6963d3/crates/ibc/src/core/context.rs#L283-L286)?